python/jiminy_py/src/jiminy_py/viewer/meshcat/meshcat_visualizer.py

```diff
@@ -37,75 +40,124 @@ def lower(self, object_data: Any) -> MsgType:
-
-    def __init__(self, height: float, radius: float):
-        super().__init__()
-        self.radius = radius
-        self.height = height
-        self.radialSegments = 32
-        self.build_triangles()
-
-    def build_triangles(self) -> None:
+    def __init__(self, height: float, radius: float, num_segments: int = 32):
         # Define proxy for convenience
-        N = self.radialSegments
-
-        # Initialize internal buffers
-        vertices, faces = [], []
+        N = num_segments
 
         # Top and bottom caps vertices
-        for e, rng in enumerate([
-                range(int(N//4) + 1), range(int(N//4), int(N//2) + 1)]):
+        vertices = []
+        for side, rng in enumerate([
+                range(int(N // 4) + 1), range(int(N // 4), int(N // 2) + 1)]):
             for i in rng:
-                for j in range(N + 1):
+                for j in range(N):
                     theta = j * 2 * math.pi / N
                     phi = math.pi * (i / (N // 2) - 1 / 2)
                     vertex = np.empty(3)
-                    vertex[0] = self.radius * math.cos(phi) * math.cos(theta)
-                    vertex[1] = self.radius * math.cos(phi) * math.sin(theta)
-                    vertex[2] = self.radius * math.sin(phi)
-                    vertex[2] += (2.0 * (e - 0.5)) * self.height / 2
+                    vertex[0] = radius * math.cos(phi) * math.cos(theta)
+                    vertex[1] = radius * math.cos(phi) * math.sin(theta)
+                    vertex[2] = radius * math.sin(phi)
+                    vertex[2] += (2.0 * (side - 0.5)) * height / 2
                     vertices.append(vertex)
+        vertices = np.vstack(vertices)
 
-        # Faces
+        # Vertex indices for faces
+        faces = []
         for i in range(int(N//2) + 1):
             for j in range(N):
-                vec = np.array([i * (N + 1) + j,
-                                i * (N + 1) + (j + 1),
-                                (i + 1) * (N + 1) + (j + 1),
-                                (i + 1) * (N + 1) + j])
-                if (i == N//4):
-                    faces.append(vec[[0, 2, 3]])
-                    faces.append(vec[[0, 1, 2]])
-                else:
-                    faces.append(vec[[0, 1, 2]])
-                    faces.append(vec[[0, 2, 3]])
-
-        # Convert to array
-        self.vertices = np.vstack(vertices).astype(np.float32)
-        self.faces = np.vstack(faces).astype(np.uint32)
-        self.normals = self.vertices
+                vec = np.array([i * N + j,
+                                i * N + (j + 1) % N,
+                                (i + 1) * N + (j + 1) % N,
+                                (i + 1) * N + j])
+                faces.append(vec[[0, 1, 2]])
+                faces.append(vec[[0, 2, 3]])
+        faces = np.vstack(faces)
+
+        # Init base class
+        super().__init__(vertices, faces)
+
+
+class DaeMeshGeometryWithTexture(ReferenceSceneElement):
```

Comment:
Could you port it back to MeshCat?
I think it can be helpful to many people beyond this project. Don't you think so?

Reply:
Yes, there is already a PR opened on the official meshcat-python repository https://github.com/rdeits/meshcat-python/pull/112. I'm getting involved but not contributing directly. I'm afraid it is going to take quite some time, mainly because the owner is quite reluctant to merge new contributions. It was more of a side project just to play around for a few hours and I'm not sure it is robust enough to be integrated in meshcat itself as is. I'm even considering dropping meshcat completely.